Add C++ Style Guide section

tiziano88 opened the pull request:
I'm kind of arbitrarily deciding some of these things but I don't really strongly feel either way, so please let me know if we should do something different. In particular about the `using` statements.

aalmos commented:
I'd warmly suggest TOTWs for pretty much everything style related. The only problem with that that's only accessible from within Google :/ and probably confidential.

BTW do you know if there exists a tool that can enforce stuff like include file order etc as suggested in the styleguide? clang-format really only formats the code. Also I was trying to get [IWYU](https://include-what-you-use.org/) working with Bazel, but no luck yet.

liulu1124 commented:
FYI a lot of TOTWs are actually public as part of open source abseil project, and it'll be good to refer to them whenever possible :)

https://abseil.io/tips/

michael-kernel-sanders commented:
https://abseil.io/tips/130

michael-kernel-sanders commented:
@aalmos clang-format does do include ordering.  For more than formatting there's also clang-tidy.

tiziano88 commented:
> I'm leaning toward just following go/totw/151 at this point.
> 
> Should probably also link to https://google.github.io/styleguide/cppguide.html.

Done, though I added details on fully qualified names since totw/151 is not available externally. I also cleaned up the code to follow the guidelines, PTAL.

michael-kernel-sanders commented on CONTRIBUTING.md:
```diff
@@ -34,6 +34,39 @@ do not expect commits to have meaningful messages during a review; the PR
+    C++ code formatted consistently.
+-   Use fully qualified names (leading `::`) for `using` declarations and
+    namespace aliases, and avoid fully qualified names for everything else,
+    unless it is necessary to make the code compile.
```
File to call this out here, but you can also link to https://abseil.io/tips/130.

tiziano88 replied:
I thought the relevant one would be 151, but that's not available externally. 130 seems more about introducing new namespaces, so I thought I'd call it out anyways.
